rabbitmq/persistent_producer.go

```diff
@@ -0,0 +1,108 @@
+package rabbitmq
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/palantir/stacktrace"
+	"github.com/streadway/amqp"
+	"go.uber.org/zap"
+
+	"github.com/sumup-oss/go-pkgs/logger"
+)
+
+type PersistentProducer struct {
+	producer          *Producer
+	mu                sync.RWMutex
+	reconnectTimeout  time.Duration
+	rabbitClientCfg   *ClientConfig
+	rabbitClientSetup *Setup
+}
+
+func NewPersistentProducer(
+	ctx context.Context,
+	client *RabbitMQClient,
+	setup *Setup,
+	logger logger.StructuredLogger,
+	metric Metric,
+	reconnectTimeout time.Duration,
+) (*PersistentProducer, error) {
+	producer, err := NewProducer(client, logger, metric)
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "failed to create a persistent producer")
+	}
+
+	persistentProducer := &PersistentProducer{
+		producer:          producer,
+		reconnectTimeout:  reconnectTimeout,
+		rabbitClientCfg:   client.cfg,
+		rabbitClientSetup: setup,
+	}
+
+	go persistentProducer.unsafeReconnect(ctx)
+
+	return persistentProducer, nil
+}
+
+func (p *PersistentProducer) Publish(
+	exchange,
+	key string,
+	mandatory,
+	immediate bool,
+	expiration string,
+	body []byte,
+	args amqp.Table,
+) error {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	err := p.producer.Publish(exchange, key, mandatory, immediate, expiration, body, args)
+	if err != nil {
+		if p.producer.isClosed {
+			return stacktrace.Propagate(err, "connection to RabbitMQ client closed")
+		}
+		return stacktrace.Propagate(err, "failed to publish RMQ message")
+	}
+
+	return nil
+}
+
+func (p *PersistentProducer) unsafeReconnect(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			p.producer.logger.Info("received shut down signal")
+			return
+		case <-p.producer.closeCh:
+			p.producer.logger.Info("try to reconnect RabbitMQ producer")
+			client, err := NewRabbitMQClient(ctx, p.rabbitClientCfg)
+			if err != nil {
+				p.producer.logger.Warn("RabbitMQ Failed to init client", zap.Error(err))
+				time.Sleep(p.reconnectTimeout)
```

Comment:
Not sure how much of an issue this is but maybe a good idea to put a select here as well (it's a nested select so it's ugly) but just being mindful that we'd block here for the duration of p.reconnectTimeout and we might want to just exit sooner in case of `<-ctx.Done()`

Reply:
Done.